src/hotspot/share/gc/z/zList.inline.hpp

```diff
@@ -28,47 +28,55 @@
 #include "utilities/debug.hpp"
 
 template <typename T>
-inline ZListNode<T>::ZListNode(ZListNode* next, ZListNode* prev) :
-    _next(next),
-    _prev(prev) {}
+inline ZListNode<T>::ZListNode() :
+    _next(this),
+    _prev(this) {}
 
 template <typename T>
-inline void ZListNode<T>::set_unused() {
-  _next = NULL;
-  _prev = NULL;
+inline ZListNode<T>::~ZListNode() {
+  verify_links_unlinked();
 }
 
 template <typename T>
-inline ZListNode<T>::ZListNode() {
-  set_unused();
+inline void ZListNode<T>::verify_links() const {
+  assert(_next->_prev == this, "Corrupt list node");
+  assert(_prev->_next == this, "Corrupt list node");
 }
 
 template <typename T>
-inline ZListNode<T>::~ZListNode() {
-  set_unused();
+inline void ZListNode<T>::verify_links_linked() const {
+  assert(_next != this, "Should be in a list");
+  assert(_prev != this, "Should be in a list");
+  verify_links();
 }
 
 template <typename T>
-inline bool ZListNode<T>::is_unused() const {
-  return _next == NULL && _prev == NULL;
+inline void ZListNode<T>::verify_links_unlinked() const {
+  assert(_next == this, "Should not be in a list");
+  assert(_prev == this, "Should not be in a list");
+  verify_links();
```

Comment:
If this node is not linked, why still "verify_links"?

Reply:
The call to `verify_links()` is a bit superfluous here, since `_next == this && _prev == this` implies `_next->_prev == this && _prev->_next == this`. I can remove it.

Reply:
Thank you.